imap_processing/utils.py

```diff
@@ -254,3 +255,86 @@ def update_epoch_to_datetime(dataset: xr.Dataset) -> xr.Dataset:
     dataset = dataset.assign_coords(epoch=epoch)
 
     return dataset
+
+
+def packet_file_to_datasets(
+    packet_file: Union[str, Path], xtce_packet_definition: str
+) -> dict[int, xr.Dataset]:
+    """
+    Convert a packet file to xarray datasets.
+
+    The packet file can contain multiple apids and these will be separated
+    into distinct datasets, one per apid. The datasets will contain the
+    ``derived_value``s of the data fields, and the ``raw_value``s if no
+    ``derived_value`` is available. If there are conversions in the XTCE
+    packet definition, the ``derived_value`` will be the converted value.
+    The dimension of the dataset will be the time field in J2000 nanoseconds.
+
+    Parameters
+    ----------
+    packet_file : str
+        Path to data packet path with filename.
+    xtce_packet_definition : str
+        Path to XTCE file with filename.
+
+    Returns
+    -------
+    datasets : dict
+        Mapping from apid to xarray dataset, one dataset per apid.
+    """
+    # Set up containers to store our data
+    # We are getting a packet file that may contain multiple apids
+    # Each apid has consistent data fields, so we want to create a
+    # dataset per apid.
+    # {apid1: dataset1, apid2: dataset2, ...}
+    data_dict: dict[int, dict] = dict()
+    attribute_dict: dict[int, dict] = dict()
+
+    # Set up the parser from the input packet definition
+    packet_definition = xtcedef.XtcePacketDefinition(xtce_packet_definition)
+    packet_parser = parser.PacketParser(packet_definition)
+
+    with open(packet_file, "rb") as binary_data:
+        packet_generator = packet_parser.generator(binary_data)
+        for packet in packet_generator:
+            apid = packet.header["PKT_APID"].raw_value
+            if apid not in data_dict:
+                # This is the first packet for this APID
+                data_dict[apid] = collections.defaultdict(list)
+                attribute_dict[apid] = collections.defaultdict(dict)
+
+                # Fill in the attributes
+                for key, value in packet.data.items():
+                    attribute_dict[apid][key]["short_description"] = (
+                        value.short_description
+                    )
+                    attribute_dict[apid][key]["long_description"] = (
+                        value.long_description
+                    )
+                    # TODO: Additional info from XTCE like units
+
+            # TODO: Include the header too? (packet.header | packet.data)
```

Comment:
Generally, l1a is close to the unprocessed packet, so I would argue that similar to raw vs. derived value, the header should be included for l1a an not included for l1b.

Reply:
OK, I'm happy to do that as well, then we could strip it later.

For the raw vs `derived_value`. Do we not want to use the XTCE functionality of `derived_value` and rather just do that ourselves in the code later? i.e. use the raw_to_eu conversion function instead.

Reply:
I added an optional argument to it that defaults to `derived_value`, but allows one to get the raw value if we want too. I can remove it or change the default if you think this would be better as `raw` by default. I actually think we should use derived if we can, but I'm not wedded to that.